### DISCUSSION TITLE
Replace elements of ToMany collection without fetching them

### DISCUSSION THREAD

**Discussion by belovaf** [Blazebit/blaze-persistence#1822] [Q&A]
I have the following model:
```
@UpdatableEntityView
@EntityView(Parent::class)
interface ParentView {
    @get:IdMapping
    val id: Long?

    @get:MappingInverse(removeStrategy = REMOVE)
    @get:UpdatableMapping(orphanRemoval = true, cascade = [PERSIST])
    val children: MutableList<ChildView>
}

@CreatableEntityView
@EntityView(Child::class)
interface ChildView {
    @get:IdMapping
    val id: Long?
    var name: String
}
```

Is it possible to fetch `ParentView` and replace all its children without fetching them?
I have tried it with `convert`, but it doesn't work with collection attributes, because they are null after conversion.

**Comment by beikov** [comment]
Hi, what have you tried so far? You should be able to use something along the lines of this:

```java
ParentView v = evm.getReference(ParentView.class, parentId);
v.setChildren( newChildren );
evm.saveFull(entityManager, v);
```

**Reply by belovaf** [reply]
I have tried:
1) Get a reference and save it (your variant). This variant can't be used with optimistic lock on parent (my case).
2) Load a lightweight updatable parent view and `convert` it to the `ParentView`.
2.1) When I set its children with setter, I get `Replacing a collection that PERSIST or UPDATE cascades is prohibited by default`. Disabling `UPDATER_STRICT_CASCADING_CHECK` leads to other weird errors.
2.2) When I use a mutable collection in the converted view I get `NPE` because field is null.

Is it possible somehow to inject a `RecordingList` to the converted view?

**Reply by beikov** [reply]
Could you please create a GitHub issue for the problem with the optimistic version? I think this should be fixed if possible.
I don't know how/if converting could be made to work, I'll have to think a bit about this.

In the meantime, you could use the collection DML API to interact with the join tables manually. See https://persistence.blazebit.com/documentation/1.6/core/manual/en_US/index.html#delete-collection-statement and https://persistence.blazebit.com/documentation/1.6/core/manual/en_US/index.html#insert-select-collection-statement

**Reply by belovaf** [reply]
Ok I will create issue, thank you!

Btw, i use a widening conversion with `IGNORE_MISSING_ATTRIBUTES` in similar cases for simple fields.
For example, if I always replace a user `fullName` and I don't need a previous `fullName`, I do the following:
1) fetch updatable view without `fullName`
2) convert view to extended view with `fullName`
3) set `fullName`
4) save converted view

Is this correct approach?

I can easily use same approach with collection fields too, but currently they are nulls after conversion.

**Reply by beikov** [reply]
Yeah, that approach sounds reasonable to me, but can you explain a bit more why you're loading a view that doesn't contain `fullName` though still want to update that field? I'd like to understand the use case a bit to see if I might be missing something or can suggest you a better alternative.

>I can easily use same approach with collection fields too, but currently they are nulls after conversion.

Ok, and when you replace the collection you're getting the previously mentioned errors I suppose?

**Reply by belovaf** [reply]
The main reason is to avoid overfetch. In practice I have 10-15 replace-only fields (including embedded json).
Yes, you are right, with collections I'm getting those errors.

**Reply by belovaf** [reply]
I think, the most convenient way to declare such fields for me is just omit getters for them.

**Reply by beikov** [reply]
>The main reason is to avoid overfetch. In practice I have 10-15 replace-only fields (including embedded json).

Can you share an example? I'm having a hard time imagining a UI that allows to replace a value only, without also showing the value.

>Yes, you are right, with collections I'm getting those errors.

Got it, then please create a GitHub issue for that and I'll try to look into it :)

**Reply by belovaf** [reply]
Yes, I have a user profile settings usecase, for example.

1) User fetches all settings with view projection. Obviously I need to fetch all data here.
2) User `PUT`s updated version of settings. I don't care about previous values here, I just want to do update with optimistic locking. I intentionally don't use dirty checking here, because partial updates can lead to very huge amount of prepared statements.

Does it make sense?

**Reply by beikov** [reply]
Ok, that does make sense, though I believe in that particular case it would be more appropriate to use the `getReference()` approach. We just need to make sure that setting the version is possible. Applying the optimistic version is worth another GitHub issue :)

**Reply by belovaf** [reply]
Yes, you are right. I use `find` + `convert` as a workaround for `getReference` + set version.

Thank you, I will create two separate issues.

**Reply by belovaf** [reply]
I have created 2 issues: [1828](https://github.com/Blazebit/blaze-persistence/issues/1828), [1829](https://github.com/Blazebit/blaze-persistence/issues/1829).
